assembly/src/v1/parsers/stack_ops.rs

```diff
@@ -38,9 +38,97 @@ pub fn parse_swapw(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), As
     unimplemented!()
 }
 
-/// TODO: implement
-pub fn parse_movup(_span_ops: &mut Vec<Operation>, _op: &Token) -> Result<(), AssemblyError> {
-    unimplemented!()
+/// Translates movup.x assembly instruction to VM operations.
+/// We specifically utilize the MovUpX VM operations for indexes that match
+/// exactly with the assembly instruction.
+/// The reamaining ones we swap the word
+/// containing the index, move the particular item to the top, swap the word back,
+/// and move the item back to the top.
+pub fn parse_movup(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+    match op.num_parts() {
+        0..=1 => return Err(AssemblyError::missing_param(op)),
+        2 => match op.parts()[1] {
+            "2" => {
+                span_ops.push(Operation::MovUp2);
+            }
+            "3" => {
+                span_ops.push(Operation::MovUp3);
+            }
+            "4" => {
+                span_ops.push(Operation::MovUp4);
+            }
+            "5" => {
+                span_ops.push(Operation::SwapW);
+                span_ops.push(Operation::Swap);
+                span_ops.push(Operation::MovUp8);
+                span_ops.push(Operation::Swap);
+                span_ops.push(Operation::MovDn8);
+                span_ops.push(Operation::MovDn3);
+                span_ops.push(Operation::SwapW);
+                span_ops.push(Operation::MovUp8);
```

Comment:
Originally I thought that this could be accomplished in just 4 cycles via: `SWAPW SWAP SWAPW MOVUP4` - but if the depth of the stack is smaller than 8, the first `SWAPW` will fail.

An alternative approach is `PAD PAD PAD MOVUP8 MOVDN3 DROP DROP DROP` but this requires 8 cycles, which is a lot.

One thing we can do is to re-define how `SWAPW` works as follows:
1. If the stack has 8 or more items on it, it works as expected.
2. If the stack has fewer than 8 items, it works as if there were 8 items on the stack and uses `0` for the missing items.

The second approach will be more efficient - but I'll need to think through its implications elsewhere.